compiler/rustc_parse/src/parser/path.rs

```diff
@@ -247,8 +247,24 @@ impl<'a> Parser<'a> {
             segments.push(segment);
 
             if self.is_import_coupler() || !self.eat_path_sep() {
-                if style == PathStyle::Expr
-                    && self.may_recover()
+                let ok_for_recovery = match style {
+                    PathStyle::Expr => self.may_recover(),
+                    PathStyle::Type => {
+                        self.may_recover()
+                            && if let Some((ident, _)) = self.prev_token.ident() {
+                                ident.as_str().chars().all(|c| c.is_lowercase())
+                                    && self.token == token::Colon
+                                    && self.token.span.lo() == self.prev_token.span.hi()
+                                    && self.look_ahead(1, |token| {
+                                        self.token.span.hi() == token.span.lo()
+                                    })
+                            } else {
+                                false
+                            }
+                    }
+                    _ => false,
+                };
```

Comment:
```suggestion
                let ok_for_recovery = self.may_recover() && match style {
                    PathStyle::Expr => true,
                    PathStyle::Type if let Some((ident, _)) = self.prev_token.ident()
                                    && self.token == token::Colon
                                    && self.token.span.lo() == self.prev_token.span.hi()
                                    && self.look_ahead(1, |token| {
                                        self.token.span.hi() == token.span.lo()
                                    })
                    }
                    _ => false,
                };
```

- Moving `self.may_recover()` out of the match & using an if-let guard for conciseness
-  Getting rid of the letter case check since I'd be nice to recover from `Self:AssocTy`, `TyParam:AssocTy`, too

Reply:
it's hard to exclude this test case:

https://github.com/chenyukang/rust/blob/293514c34acd62fd71b3fee5889b414d466a3f64/tests/ui/array-slice-vec/vec-matching-fold.rs#L9

the `T:Debug` actually misses a whitespace, after formatting it's `T: Debug`.

Reply:
maybe we can ignore this scenario? since most of the time we are parsing formatted code ...